### DISCUSSION TITLE
Aborting custom script runs

### DISCUSSION THREAD

**Discussion by agreusel** [netbox-community/netbox#10430] [Q&A] [answered]
I'm trying to figure out how to terminate a custom script mid-run (via the GUI) under specified conditions.

I've tried using "sys.exit()" and "raise SystemExit", but both just change the script status to "running" and then sit there forever. The only successful "method" I can find is to intentionally run a bad command to cause an error.

Is there a cleaner way to abort a script mid-run?

**Comment by kkthxbye-code** [comment, marked as answer]
Just raise any exception.

**Reply by agreusel** [reply]
I've tried that. The relevant code is here:

```
        try:
            virtual_chassis = VirtualChassis.objects.get(name = f"{site_lower}-core01")
            self.log_failure(f"Virtual Chassis {virtual_chassis} exists, aborting.")
            raise NameError

        except VirtualChassis.DoesNotExist:
            pass
```

I know the try statement is working because I get the log_failure message. But the exception is not being raised, and the script eventually errors out in the next code block where I'm creating a new virtual chassis with the same name as an existing one. The exception I'm trying to raise never gets processed/logged.

**Reply by kkthxbye-code** [reply]
Result of the following code:

```
from extras.scripts import Script
from dcim.models import VirtualChassis

class Script0(Script):
    def run(self, data, commit):
        try:
            virtual_chassis = VirtualChassis.objects.get(name = "testvc")
            self.log_failure(f"Virtual Chassis {virtual_chassis} exists, aborting.")
            raise NameError
        except VirtualChassis.DoesNotExist:
            pass
```

Which results in:
![image](https://user-images.githubusercontent.com/400797/191574551-abb8a19a-4445-40f6-92d6-41ef33100cc9.png)

So it works exactly as it should. So either you are doing something wrong or you didn't post all the relevant code. Are you sure you don't have another try block around it or another except after the one in the snippet.

**Reply by agreusel** [reply]
I'm certain there's no other try block or exception, this is the first action in the run function. The additional code looks like:
```
    def run(self, data, commit):
        """ Create all Netbox configuration for core infrastructure """

        ### Get site code
        site = data['site']
        site_lower = site.slug
        site_upper = site.slug.upper()

        ### Create Virtual Chassis for Core Switch
        try:
            virtual_chassis = VirtualChassis.objects.get(name = f"{site_lower}-core01")
            self.log_failure(f"Virtual Chassis {virtual_chassis} exists, aborting.")
            raise NameError

        except VirtualChassis.DoesNotExist:
            pass

        virtual_chassis = VirtualChassis(
            name = f"{site_lower}-core01",
            domain = '100'
        )
        virtual_chassis.save()
        self.log_success(f"Created new virtual chassis {virtual_chassis}")
```
This results in:
![Screen Shot 2022-09-21 at 10 59 16 AM](https://user-images.githubusercontent.com/19214158/191577261-76730fe2-b459-429f-8c6b-35b00d3651b8.png)

The only other contextual data I can provide is that this is NetBox v3.2.7.

**Reply by kkthxbye-code** [reply]
Just tested in v3.2.7 and it still works. Using your script verbatim.

Maybe it's your python version, I can't remember if there's been any changes to exceptions recently. Maybe try to raise with a message, that's more common.

`raise NameError("VC Exists")`

**Reply by agreusel** [reply]
This is so weird, even a NameError with a message doesn't work.

Regardless, this definitely seems like an issue with my specific env (probably python) and not NetBox. I'll go ahead and mark this closed. Appreciate you looking at it!

**Comment by hagbarddenstore** [comment]
```
if VirtualChassis.objects.filter(name=f'{site_lower}-core01').exists():
    return
```

Simply return within your run function.

**Reply by kkthxbye-code** [reply]
Works if you have made no database changes before that point. Normally when you want to abort, you want to trigger the transaction rollback, which is not achieved by returning. It will also show the script as succeding, which is not really what I would expect the status as when aborting.
